content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+
+---
+
+## GDB and Coredump
+
+We can start with trying to inspect the running service with GDB (GNU Debugger).
+
+*Running a debugger in the production environment will likely require additional privileges. If in doubt, be wise and always consult with your operations team first.*
+
+Connect to another SSH session on the VM, find server process's id, and attach to the process with the debugger:
+
+```
+= vagrant ssh server-test-1
+Welcome to Ubuntu 18.04.1 LTS (GNU/Linux 4.15.0-33-generic x86_64)
+
+:~$ ps -ef | grep server
+vagrant   1628  1557  0 14:45 pts/0	00:00:01 ./server --addr=:10080
+
+:~$ sudo gdb --pid=1628 /vagrant/example/server/server
+```
+
+With the debbuger attached to the process, we can run GDB's `bt` command (aka backtrace) to check the stack trace of the current thread:
+
+```
+(gdb) bt
+#0  runtime.futex () at /usr/local/go/src/runtime/sys_linux_amd64.s:532
+#1  0x000000000042b08b in runtime.futexsleep (addr=0xa9a160 <runtime.m0+320>, ns=-1, val=0) at /usr/local/go/src/runtime/os_linux.go:46
+#2  0x000000000040c382 in runtime.notesleep (n=0xa9a160 <runtime.m0+320>) at /usr/local/go/src/runtime/lock_futex.go:151
+#3  0x0000000000433b4a in runtime.stoplockedm () at /usr/local/go/src/runtime/proc.go:2165
+#4  0x0000000000435279 in runtime.schedule () at /usr/local/go/src/runtime/proc.go:2565
+#5  0x00000000004353fe in runtime.park_m (gp=0xc000066d80) at /usr/local/go/src/runtime/proc.go:2676
+#6  0x000000000045ae1b in runtime.mcall () at /usr/local/go/src/runtime/asm_amd64.s:299
+#7  0x000000000045ad39 in runtime.rt0_go () at /usr/local/go/src/runtime/asm_amd64.s:201
+#8  0x0000000000000000 in ?? ()
+```
+
+Honestly, I’m not a GDB expert, but it seems that Go runtime is putting threads to sleep. *But why?*
+
+Debugging a live process is “fun”, also let’s grab a coredump of the process to analyse it offline. We can do it with GDB's `gcore` command. The core file will be saved as `core.<process_id>` in the current working directory. Note, even for our simple server, the file will be pretty big. It’s likely to be huge for production service.
+
+```
+(gdb) gcore
+Saved corefile core.1628
+(gdb) quit
+
+= du -h core.1628
+1.2G    core.1628
+```
+
+With coredump saved, we're no longer required to keep the process running. Feel free to “kill -9” it.
+
+*For more info on the topic of debugging with GDB, check out Go’s own "[Debugging Go Code with GDB][2]".*
+
+## Enter Delve, Debugger for Go
+
+[Delve][] is the debugger for Go programs. It is similar to GDB but is aware of Go's runtime, data structures and other internals of the language.
+
+I highly recommend a talk by Alessandro Arzilli “[Internal Architecture of Delve, a Debugger For Go](https://www.youtube.com/watch?v=IKnTr7Zms1k)” from GopherCon EU 2018 if you're interested to know about Delve's internals.
+
+Delve is written in Go so installing it as simple as running:
+
+```
+= go get -u github.com/derekparker/delve/cmd/dlv
```

Comment:
Where do you execute this command? Inside the VM or outside? It's clear to me now, but readers may not know.